### ISSUE TITLE
Excessive "modified memory" usage on Windows Server 2008 R2

### ISSUE THREAD

**Issue by alastrange** [elastic/topbeat#98]
We tried evaluating topbeat-1.0.0-beta4-windows on a cluster on Windows Server 2008 R2 machines, and after a few days they all had issues with "Modified Memory" usage.  One of the 64GB machines had basically run out of physical memory.  In Task Manager, none of the processes showed excessive memory usage, but the topbeat exe had a ton of page faults.  When looking at Resource Monitor was it apparent *something* was using a massive chunk of **Modified Memory**, but no indication of what was using it.  I supposed it was topbeat since that was the only new addition, and once killed all the memory was restored.

As for the layout, topbeat was shipping logs directly to a single elastic search node that didn't appear to have any disruptions during this time.

**Comment by andrewkroh** [comment]
@alastrange Thanks for reporting the issue. I am looking into this.

Here is a look at the heap after running for ~45 minutes with `period: 1`. It looks like the [`procsMap`](https://github.com/elastic/topbeat/blob/master/topbeat.go#L182) is a leak. It gets written to, but there are no `delete`'s. If there are a lot of processes being created this could be bad.

```
Fetching profile from http://localhost:6060/debug/pprof/heap
Saved profile in \pprof\pprof.localhost:6060.inuse_objects.inuse_space.005.pb.gz
Entering interactive mode (type "help" for commands)
(pprof) top10
34310.83kB of 34310.83kB total (  100%)
Dropped 206 nodes (cum <= 171.55kB)
Showing top 10 nodes out of 37 (cum >= 512.04kB)
      flat  flat%   sum%        cum   cum%
25605.91kB 74.63% 74.63% 32262.26kB 94.03%  main.(*Topbeat).exportProcStats
 3584.25kB 10.45% 85.08%  3584.25kB 10.45%  github.com/elastic/gosigar.CarrayToString
 2560.09kB  7.46% 92.54%  6656.35kB 19.40%  main.GetProcess
  512.31kB  1.49% 94.03%   512.31kB  1.49%  runtime.makeslice
  512.19kB  1.49% 95.52%   512.19kB  1.49%  runtime.malg
  512.04kB  1.49% 97.02%   512.04kB  1.49%  encoding/json.newStructEncoder
  512.03kB  1.49% 98.51%   512.03kB  1.49%  main.(*Topbeat).exportSystemStats
  512.01kB  1.49%   100%   512.01kB  1.49%  time.Time.Format
         0     0%   100%   512.04kB  1.49%  encoding/json.(*encodeState).marshal
         0     0%   100%   512.04kB  1.49%  encoding/json.(*encodeState).reflectValue
```

I think the bigger problem is the rate of leaked handles and the number of page faults. I see one missing `close` for the handle returned by [`GetParentPid`](https://github.com/elastic/topbeat/blob/master/Godeps/_workspace/src/github.com/elastic/gosigar/sigar_windows.go#L320). I will do some more testing next week.

![topbeat-handle-leak](https://cloud.githubusercontent.com/assets/4565752/11010442/da12f780-84ac-11e5-95c6-a7bdddc518f9.png)

**Comment by andrewkroh** [comment]
I did more testing today and didn't find any additional resource leaks. I was able to verify that the following changes fix the problem: https://github.com/andrewkroh/topbeat/commit/dedde2a0abcb5ccc591c3fa79fcaa41fc36631a1. I will open pull requests to fix the issues.

Also it is important to note that one of the leaks I found affects all operating systems. This graph shows the symptoms of the leak on linux.

![screen shot 2015-11-09 at 9 20 39 am](https://cloud.githubusercontent.com/assets/4565752/11049236/034ef6b6-870b-11e5-8658-80f6f47b3227.png)
